### DISCUSSION TITLE
GTC 2021 talks

### DISCUSSION THREAD

**Discussion by hwu36** [NVIDIA/cutlass#223] [General]
There will be at least 3 CUTLASS related talks in the coming GTC21.

First, @manishucsd  will dive into the details of CUTLASS convolution kernels in https://gtc21.event.nvidia.com/media/Accelerating%20Convolution%20with%20Tensor%20Cores%20in%20CUTLASS%20%5BS31883%5D/1_ofvh631z (Apr 13, 2021 | 4:00 PM - 4:40 PM EDT).

Second, Paul Eller from Oak Ridge National Laboratory will introduce how CUTLASS 1-bit tensor core implementation is used to accelerate large-scale genomic applications in https://gtc21.event.nvidia.com/media/Parallel%20Accelerated%201-bit%20Dense%20Matrix%20Multiplies%20for%20Genomics%20Applications%20%5BS31560%5D/1_5em1atq3 (Apr 14, 2021 | 4:00 PM - 4:20 PM EDT)

Last, I will answer any CUTLASS questions in a Connect with the Experts session together with other maths library representatives in https://gtc21.event.nvidia.com/media/Tensor%20Core-Accelerated%20Math%20Libraries%20for%20Dense%20and%20Sparse%20Linear%20Algebra%20in%20AI%20and%20HPC%20%5BCWES1098%5D/1_z6ept1ky (Apr 19, 2021 | 12:00 PM - 1:00 PM EDT)

CUTLASS will also be mentioned by several other NVIDIA talks.  Enjoy GTC 2021!

(Please leave us a star in our github if you haven't done so)

**Comment by hwu36** [comment]
For @manishucsd 's Tuesday Convolution talk, @jwang323 and I will also answer questions in the backstage.  Don't miss the talk.

**Reply by illava** [reply]
Do we have a replay or summarize of those answered questions?

**Reply by hwu36** [reply]
no, I don't think GTC is going to release the QA part.

**Reply by hwu36** [reply]
I remember one question is similar to #231

**Comment by hwu36** [comment]
And don't miss GTC keynote on Monday.

**Comment by hwu36** [comment]
In my connect with the experts session, @Peter9606 asked a question "To cublas, does cublas use cutlass ? If the answer is yes, which level ? warp, or CTA or kernel ?"  Watch the [answer](https://gtc21.event.nvidia.com/media/t/1_0znia7s8) from 31:50 .

Another question raised in the session is "When to use libraries and when to use cutlass?"  My answer would be
- If you just need an api, libraries such as cublas and cudnn are worry-free one-stop shops. 
- You would need cutlass, when
1. You need source code to customize,  or to just use some cutlass sub-modules, or to piecemeal cutlass for a DL/HPC compiler, etc.
2. Libraries such as cuDNN/cuBLAS may not pick the best kernels for your problem sizes, existing cutlass kernels or customized cutlass kernels can outperform the libraries in these cases.
3. cutlass supports kernels that do not exist in the libraries, such as int4, int1 kernels, and complex conv, etc. 
4. Libraries like cudnn/cublas are huge and > 99.9% of their kernels are not needed by a typical user.  in cutlass, you just need to instantiate what you need.
5. libraries like cudnn/cublas has CPU overhead.  This overhead can be minimum in cutlass if you know what kernel you want.  It is important for small problem size.

Another question is that how to start with cutlass?  My answer would be

- Our [quickstart](https://github.com/NVIDIA/cutlass/blob/master/media/docs/quickstart.md) introduces how to clone, build, run unit tests.
- Our [profiler](https://github.com/NVIDIA/cutlass/blob/master/media/docs/profiler.md) introduces how to use our profiler to test the performance of the gemm/conv2d/conv3d kernels.
- Former GTC talks helps understand the basics of writing CUTLASS kernels: https://developer.nvidia.com/gtc/2020/video/s21745-vid and https://on-demand.gputechconf.com/gtc/2018/presentation/s8854-cutlass-software-primitives[…]-linear-algebra-at-all-levels-and-scales-within-cuda.pdf
